src/bitcoin-util.cpp

```diff
@@ -0,0 +1,217 @@
+    CBlockHeader header = header_orig; // working copy
+    header.nNonce = offset;
+
+    uint32_t finish = std::numeric_limits<uint32_t>::max() - step;
+    finish = finish - (finish % step) + offset;
+
+    while (!found && header.nNonce < finish) {
+        const uint32_t next = (finish - header.nNonce < 5000*step) ? finish : header.nNonce + 5000*step;
+        do {
+            if (UintToArith256(header.GetHash()) <= target) {
+                if (!found.exchange(true)) {
+                    header_orig.nNonce = header.nNonce;
+                }
+                return;
+            }
+            header.nNonce += step;
+        } while(header.nNonce != next);
+    }
+}
+
+static int Grind(int argc, char* argv[], std::string& strPrint)
+{
+    if (argc != 1) {
+        strPrint = "Must specify block header to grind";
+        return 1;
+    }
+
+    CBlockHeader header;
+    if (!DecodeHexBlockHeader(header, argv[0])) {
+        strPrint = "Could not decode block header";
+        return 1;
+    }
+
+    uint32_t nBits = header.nBits;
+    std::atomic<bool> found{false};
+
+    std::vector<std::thread> threads;
+    int n_tasks = std::max(1u, std::thread::hardware_concurrency());
+    for (int i = 0; i < n_tasks; ++i) {
+        threads.emplace_back( grind_task, nBits, std::ref(header), i, n_tasks, std::ref(found) );
+    }
+    for (auto& t : threads) {
+        t.join();
+    }
+    if (!found) {
+        strPrint = "Could not satisfy difficulty target";
+        return 1;
+    }
+
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << header;
+    strPrint = HexStr(ss);
+    return 0;
+}
+
+static int CommandLineUtil(int argc, char* argv[])
+{
+    if (argc <= 1) return 1;
+
+    std::string strPrint;
+    int nRet = 0;
+
+    try {
+        while (argc > 1 && IsSwitchChar(argv[1][0]) && (argv[1][1] != 0)) {
+            --argc;
+            ++argv;
+        }
+
+        char* command = argv[1];
+        if (strcmp(command, "grind") == 0) {
+            nRet = Grind(argc-2, argv+2, strPrint);
+        } else {
+            strPrint = strprintf("Unknown command %s", command);
+            nRet = 1;
+        }
+    }
+    catch (const std::exception& e) {
+        strPrint = std::string("error: ") + e.what();
+        nRet = EXIT_FAILURE;
+    }
+    catch (...) {
+        PrintExceptionContinue(nullptr, "CommandLineUtil()");
+        throw;
+    }
+
+    if (strPrint != "") {
+        tfm::format(nRet == 0 ? std::cout : std::cerr, "%s\n", strPrint);
+    }
+    return nRet;
+}
+
+int main(int argc, char* argv[])
+{
+    if (argc < 0) {
+        /* this nonsense is needed or you get linker errors because
+         * ArgsManager does these ops but for some reason they're not
+         * available.
+         * "argc < 0" should be enough to be unreachable, but not
+         * so unreachable that it's optimised away entirely
+         */
+        try { UniValue uv; uv["txid"]; uv.get_str(); uv.write(4); } catch (...) { }
```

Comment:
I think this is awful. Can someone of the build system people please weigh in here?

Reply:
This works for me:
```diff
diff --git a/src/Makefile.am b/src/Makefile.am
index b974bed66..f37db7bde 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -677,9 +677,9 @@ bitcoin_util_SOURCES += bitcoin-util-res.rc
 endif
 
 bitcoin_util_LDADD = \
-  $(LIBUNIVALUE) \
   $(LIBBITCOIN_COMMON) \
   $(LIBBITCOIN_UTIL) \
+  $(LIBUNIVALUE) \
   $(LIBBITCOIN_CONSENSUS) \
   $(LIBBITCOIN_CRYPTO) \
   $(LIBSECP256K1)
diff --git a/src/bitcoin-util.cpp b/src/bitcoin-util.cpp
index ff6bedec0..f7e670f4e 100644
--- a/src/bitcoin-util.cpp
+++ b/src/bitcoin-util.cpp
@@ -179,16 +179,6 @@ static int CommandLineUtil(int argc, char* argv[])
 
 int main(int argc, char* argv[])
 {
-    if (argc < 0) {
-        /* this nonsense is needed or you get linker errors because
-         * ArgsManager does these ops but for some reason they're not
-         * available.
-         * "argc < 0" should be enough to be unreachable, but not
-         * so unreachable that it's optimised away entirely
-         */
-        try { UniValue uv; uv["txid"]; uv.get_str(); uv.write(4); } catch (...) { }
-    }
-
     SetupEnvironment();
 
     try {
```

Reply:
Thanks @fanquake !